File: sdk/src/main/java/eu/darken/porter/sdk/ShizukuProtocol.java
package eu.darken.porter.sdk;

/** The wire a Shizuku server speaks: descriptors, raw codes and Bundle keys. */
final class ShizukuProtocol {

    private ShizukuProtocol() {
    }

    /**
     * Written by hand rather than taken from a generated stub: the modules holding the Shizuku AIDL
     * and its constants are not published, so nothing an integrator resolves would carry them.
     */
    static final String DESCRIPTOR = "moe.shizuku.server.IShizukuService";
    static final String APPLICATION_DESCRIPTOR = "moe.shizuku.server.IShizukuApplication";
    static final String SERVICE_CONNECTION_DESCRIPTOR = "moe.shizuku.server.IShizukuServiceConnection";

    /** What attach declares this client to be. */
    static final int CLIENT_API_VERSION = 13;
    /** A server below this is refused rather than spoken to on an older encoding. */
    static final int MINIMUM_VERSION = 13;

    // Codes on DESCRIPTOR. The trailing comment is the id the method declares in the AIDL; the
    // generated stub numbers it FIRST_CALL_TRANSACTION above that, and FIRST_CALL_TRANSACTION is 1.
    /** Outside the AIDL, and answered ahead of the generated dispatch. */
    static final int TRANSACTION_transactRemote = 1;
    static final int TRANSACTION_getUid = 4; // 3
    static final int TRANSACTION_checkPermission = 5; // 4
    static final int TRANSACTION_getSELinuxContext = 9; // 8
    static final int TRANSACTION_getSystemProperty = 10; // 9
    static final int TRANSACTION_setSystemProperty = 11; // 10
    static final int TRANSACTION_addUserService = 12; // 11
    static final int TRANSACTION_removeUserService = 13; // 12
    static final int TRANSACTION_requestPermission = 15; // 14
    static final int TRANSACTION_checkSelfPermission = 16; // 15
    static final int TRANSACTION_shouldShowRequestPermissionRationale = 17; // 16
    static final int TRANSACTION_attachApplication = 18; // 17
    static final int TRANSACTION_exit = 101; // 100
    static final int TRANSACTION_attachUserService = 102; // 101
    static final int TRANSACTION_dispatchPermissionConfirmationResult = 105; // 104
    static final int TRANSACTION_getFlagsForUid = 106; // 105
    static final int TRANSACTION_updateFlagsForUid = 107; // 106

    // Codes on APPLICATION_DESCRIPTOR, which the server transacts on the client.
    static final int APPLICATION_TRANSACTION_bindApplication = 2; // 1
    static final int APPLICATION_TRANSACTION_dispatchRequestPermissionResult = 3; // 2

    // Codes on SERVICE_CONNECTION_DESCRIPTOR, which the server transacts on the client.
    static final int SERVICE_CONNECTION_TRANSACTION_connected = 1; // 0
    static final int SERVICE_CONNECTION_TRANSACTION_died = 2; // 1

    // attach args
    static final String ATTACH_APPLICATION_PACKAGE_NAME = "shizuku:attach-package-name";
    static final String ATTACH_APPLICATION_API_VERSION = "shizuku:attach-api-version";

    // bind application, which is where the attach reply arrives
    static final String BIND_APPLICATION_SERVER_UID = "shizuku:attach-reply-uid";
    static final String BIND_APPLICATION_SERVER_VERSION = "shizuku:attach-reply-version";
    static final String BIND_APPLICATION_SERVER_PATCH_VERSION = "shizuku:attach-reply-patch-version";
    static final String BIND_APPLICATION_SERVER_SECONTEXT = "shizuku:attach-reply-secontext";
    static final String BIND_APPLICATION_PERMISSION_GRANTED = "shizuku:attach-reply-permission-granted";
    static final String BIND_APPLICATION_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE =
            "shizuku:attach-reply-should-show-request-permission-rationale";

    // permission result, and the confirmation a manager sends back
    static final String REQUEST_PERMISSION_REPLY_ALLOWED = "shizuku:request-permission-reply-allowed";
    static final String REQUEST_PERMISSION_REPLY_IS_ONETIME = "shizuku:request-permission-reply-is-onetime";

    // user service
    static final String USER_SERVICE_ARG_TAG = "shizuku:user-service-arg-tag";
    static final String USER_SERVICE_ARG_COMPONENT = "shizuku:user-service-arg-component";
    static final String USER_SERVICE_ARG_DEBUGGABLE = "shizuku:user-service-arg-debuggable";
    static final String USER_SERVICE_ARG_VERSION_CODE = "shizuku:user-service-arg-version-code";
    static final String USER_SERVICE_ARG_PROCESS_NAME = "shizuku:user-service-arg-process-name";
    static final String USER_SERVICE_ARG_NO_CREATE = "shizuku:user-service-arg-no-create";
    static final String USER_SERVICE_ARG_DAEMON = "shizuku:user-service-arg-daemon";
    static final String USER_SERVICE_ARG_USE_32_BIT_APP_PROCESS = "shizuku:user-service-arg-use-32-bit-app-process";
    /** Spelled without the "-arg-" of the other eight; {@code ShizukuProtocolConstantsTest} pins it. */
    static final String USER_SERVICE_ARG_REMOVE = "shizuku:user-service-remove";
    static final String USER_SERVICE_ARG_TOKEN = "shizuku:user-service-arg-token";
}

File: sdk/src/main/java/eu/darken/porter/sdk/ShizukuProtocol.kt
package eu.darken.porter.sdk

/** The wire a Shizuku server speaks: descriptors, raw codes and Bundle keys. */
internal object ShizukuProtocol {

    /**
     * Written by hand rather than taken from a generated stub: the modules holding the Shizuku AIDL
     * and its constants are not published, so nothing an integrator resolves would carry them.
     */
    const val DESCRIPTOR: String = "moe.shizuku.server.IShizukuService"
    const val APPLICATION_DESCRIPTOR: String = "moe.shizuku.server.IShizukuApplication"
    const val SERVICE_CONNECTION_DESCRIPTOR: String = "moe.shizuku.server.IShizukuServiceConnection"

    /** The canonical manager, and the permission it declares. */
    const val MANAGER_APPLICATION_ID: String = "moe.shizuku.privileged.api"
    const val PERMISSION: String = "moe.shizuku.manager.permission.API_V23"

    /**
     * Shizuku+'s Plus flavor, which declares this permission instead of [PERMISSION] and delivers
     * its binder the same way.
     */
    const val PLUS_MANAGER_APPLICATION_ID: String = "af.shizuku.plus.api"
    const val PLUS_PERMISSION: String = "af.shizuku.plus.permission.API_V23"

    /** What attach declares this client to be. */
    const val CLIENT_API_VERSION: Int = 13

    /** A server below this is refused rather than spoken to on an older encoding. */
    const val MINIMUM_VERSION: Int = 13

    // Codes on DESCRIPTOR. The trailing comment is the id the method declares in the AIDL; the
    // generated stub numbers it FIRST_CALL_TRANSACTION above that, and FIRST_CALL_TRANSACTION is 1.
    /** Outside the AIDL, and answered ahead of the generated dispatch. */
    const val TRANSACTION_transactRemote: Int = 1
    const val TRANSACTION_getUid: Int = 4 // 3
    const val TRANSACTION_checkPermission: Int = 5 // 4
    const val TRANSACTION_getSELinuxContext: Int = 9 // 8
    const val TRANSACTION_getSystemProperty: Int = 10 // 9
    const val TRANSACTION_setSystemProperty: Int = 11 // 10
    const val TRANSACTION_addUserService: Int = 12 // 11
    const val TRANSACTION_removeUserService: Int = 13 // 12
    const val TRANSACTION_requestPermission: Int = 15 // 14
    const val TRANSACTION_checkSelfPermission: Int = 16 // 15
    const val TRANSACTION_shouldShowRequestPermissionRationale: Int = 17 // 16
    const val TRANSACTION_attachApplication: Int = 18 // 17

    // Codes on APPLICATION_DESCRIPTOR, which the server transacts on the client.
    const val APPLICATION_TRANSACTION_bindApplication: Int = 2 // 1
    const val APPLICATION_TRANSACTION_dispatchRequestPermissionResult: Int = 3 // 2

    // Codes on SERVICE_CONNECTION_DESCRIPTOR, which the server transacts on the client.
    const val SERVICE_CONNECTION_TRANSACTION_connected: Int = 1 // 0
    const val SERVICE_CONNECTION_TRANSACTION_died: Int = 2 // 1

    // attach args
    const val ATTACH_APPLICATION_PACKAGE_NAME: String = "shizuku:attach-package-name"
    const val ATTACH_APPLICATION_API_VERSION: String = "shizuku:attach-api-version"

    // bind application, which is where the attach reply arrives
    const val BIND_APPLICATION_SERVER_UID: String = "shizuku:attach-reply-uid"
    const val BIND_APPLICATION_SERVER_VERSION: String = "shizuku:attach-reply-version"
    const val BIND_APPLICATION_SERVER_PATCH_VERSION: String = "shizuku:attach-reply-patch-version"
    const val BIND_APPLICATION_SERVER_SECONTEXT: String = "shizuku:attach-reply-secontext"
    const val BIND_APPLICATION_PERMISSION_GRANTED: String = "shizuku:attach-reply-permission-granted"
    const val BIND_APPLICATION_SHOULD_SHOW_REQUEST_PERMISSION_RATIONALE: String =
        "shizuku:attach-reply-should-show-request-permission-rationale"

    // permission result
    const val REQUEST_PERMISSION_REPLY_ALLOWED: String = "shizuku:request-permission-reply-allowed"

    // user service
    const val USER_SERVICE_ARG_TAG: String = "shizuku:user-service-arg-tag"
    const val USER_SERVICE_ARG_COMPONENT: String = "shizuku:user-service-arg-component"
    const val USER_SERVICE_ARG_DEBUGGABLE: String = "shizuku:user-service-arg-debuggable"
    const val USER_SERVICE_ARG_VERSION_CODE: String = "shizuku:user-service-arg-version-code"
    const val USER_SERVICE_ARG_PROCESS_NAME: String = "shizuku:user-service-arg-process-name"
    const val USER_SERVICE_ARG_NO_CREATE: String = "shizuku:user-service-arg-no-create"
    const val USER_SERVICE_ARG_DAEMON: String = "shizuku:user-service-arg-daemon"
    const val USER_SERVICE_ARG_USE_32_BIT_APP_PROCESS: String = "shizuku:user-service-arg-use-32-bit-app-process"

    /** Spelled without the "-arg-" of the other eight; `ShizukuProtocolConstantsTest` pins it. */
    const val USER_SERVICE_ARG_REMOVE: String = "shizuku:user-service-remove"
}
